### PR TITLE
Refactor reference guide

### PR DESCRIPTION
**Pull request by salsakran**
This code is kind of a tire-fire that no one can really extend easily. The code is far too abstracted, and dumps all of its complexity in a few master components that need to know and do far too much. 

Aggressively refactoring this in hopes of eventually changing it to use the metadata components in #5134

### CODE REVIEW

**Comment by salsakran** [comment]
Ok, this is largely where I wanted the first big refactor to stop. Main thing that was accomplished was unwinding the generative approach and there is not a container + a display component for each route.

Think there needs to be another pass at this to -
1. Split each display component into a read component and an edit form component
2. Unify the loading/updating actions in reference.js with the new metadata library + redux approaches in the rest of the application

Separately, we *really* need to stop loading quite so much for the guide display. It's entirely unreasonable for a top nav-level click to result in loading every dashboard, every metric, every segment and every database/table/field metadata.

File: frontend/src/metabase/reference/databases/TableSidebar.jsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import pure from "recompose/pure";
+
+const TableSidebar = ({
+    database,
+    table,
+    style,
+    className
+}) =>
+    <div className={cx(S.sidebar, className)} style={style}>
+        <ul>
```
Comment (kdoh) [inline frontend/src/metabase/reference/databases/TableSidebar.jsx]: Bit pedantic but I'd change this from an `<ul>` to an `<ol>` since the order of the items in the sidebar is significant. I'd also scoot that down to wrap the side bar items since a list should have list items as its immediate children.

File: frontend/src/metabase/reference/databases/TableSidebar.jsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                             icon="document" 
+                             name="Details" />
+                <SidebarItem key={`/reference/databases/${database.id}/tables/${table.id}/fields`} 
+                             href={`/reference/databases/${database.id}/tables/${table.id}/fields`} 
+                             icon="fields" 
+                             name="Fields in this table" />
+                <SidebarItem key={`/reference/databases/${database.id}/tables/${table.id}/questions`} 
+                             href={`/reference/databases/${database.id}/tables/${table.id}/questions`} 
+                             icon="all" 
+                             name={`Questions about this table`} />
```
Comment (kdoh) [inline frontend/src/metabase/reference/databases/TableSidebar.jsx]: Small thing here but since there's no dynamic data here you can just use normal quotes and remove the need for braces .

File: frontend/src/metabase/reference/guide/BaseSidebar.jsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import cx from 'classnames';
+import pure from "recompose/pure";
+
+const BaseSidebar = ({
+    style,
+    className
+}) =>
+    <div className={cx(S.sidebar, className)} style={style}>
+        <ul>
```
Comment (kdoh) [inline frontend/src/metabase/reference/guide/BaseSidebar.jsx]: Same deal here about ul -> ol and the placement.

**Review by kdoh — COMMENTED** [review]
Much, much easier to parse so big props for that.

Overall the changes seem good. I left one pedantic comment about the list structure which applies to all the new side bars.

There are also some general indentation / quotes things that I didn't want to make comments on for each one. Might be worth running `prettier` over the stuff in the reference directory just to clean all that up.

I'll take another look once more tests are there.

File: frontend/src/metabase/reference/databases/DatabaseDetail.jsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    }
+};
+
+const mapDispatchToProps = {
+    ...metadataActions,
+    ...actions,
+    onChangeLocation: push
+};
+
+const validate = (values, props) => {
```
Comment (kdoh) [inline frontend/src/metabase/reference/databases/DatabaseDetail.jsx]: This seems like it doesn't really validate anything?

**Review by attekei — COMMENTED** [review]
I took a pretty quick look at this now because I was struggling with changing the "interesting questions" logic in metabase-lib branch.

Routes.jsx looks way more sensible than before and I also see a clear path for getting rid of the inflexible `getQuestionUrl` function which blocks my refactoring efforts.

Will take a more careful look + test manually the functionality after some integration tests are in place.

**Review by kdoh — APPROVED** [review]
Nice work on this. Once tests are all good this should be g2g from my perspective.

**Comment by attekei** [comment]
I did a tiny bug bash session for the data reference section, here are my findings:

### Can't change foreign key when editing a table
screen shot 2017-07-07 at 14 55 45

### Entering a field through a metric/segment fails if you come directly through the url
screen shot 2017-07-07 at 14 52 39

### Clicking "Edit" for many kind of entities doesn't prefill the existing entity information

screen shot 2017-07-07 at 14 59 24

screen shot 2017-07-07 at 14 59 30

There is also a cluster of bugs I've already fixed in query lib branch, so these probably don't need to be addressed here:
- Clicking "View the metric formula" or "View the segment formula" often shows malformed formula
- "Potentially useful questions" links in fields/segments/metrics don't always work correctly (for example grouping by a date field)

**Comment by salsakran** [comment]
good catches

File: frontend/src/metabase/reference/components/EditableReferenceHeader.jsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            </div>
+            { type === 'table' && !hasSingleSchema && !isEditing &&
+                <div className={S.headerSchema}>{entity.schema}</div>
+            }
+            <div
+                className={S.headerBody}
+                style={isEditing && name === 'Details' ? {alignItems: "flex-start"} : {}}
+            >
+                { isEditing && name === 'Details' ?
+                    hasDisplayName ?
```
Comment (attekei) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: this kind of nested ternary operators are hard to follow, could the inner ternary operator be at least wrapped in parentheses? (this isn't a blocking issue)

Comment (attekei) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: probably this code wasn't originally written by you but I still comment if I encounter stuff that is really convoluted

File: frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+export default class DatabaseDetailContainer extends Component {
+    static propTypes = {
+        params: PropTypes.object.isRequired,
+        location: PropTypes.object.isRequired,
+        database: PropTypes.object.isRequired,
+        isEditing: PropTypes.bool
+    };
+
+    async fetchContainerData(){
+        await actions.rFetchDatabaseMetadata(this.props, this.props.databaseId);
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: Why this is `rFetch`, not `fetch`? If `r` means something, then why not writing the whole word there

File: frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    async componentWillReceiveProps(newProps) {
+        if (this.props.location.pathname === newProps.location.pathname) {
+            return;
+        }
+
+        newProps.endEditing();
+        newProps.endLoading();
+        newProps.clearError();
+        newProps.collapseFormula();
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: this cluster of action invocations is kinda ugly

File: frontend/src/metabase/reference/databases/DatabaseList.jsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    getDatabases,
+    getError,
+    getLoading
+} from "../selectors";
+
+import * as metadataActions from "metabase/redux/metadata";
+
+const emptyStateData = {
+            title: "Metabase is no fun without any data",
+            adminMessage: "Your databses will appear here once you connect one",
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/DatabaseList.jsx]: typo "databses"

File: frontend/src/metabase/reference/databases/DatabaseListContainer.jsx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+@connect(mapStateToProps, mapDispatchToProps)
+export default class DatabaseListContainer extends Component {
+    static propTypes = {
+        params: PropTypes.object.isRequired,
+        location: PropTypes.object.isRequired,
+        isEditing: PropTypes.bool
+    };
+
+    async fetchContainerData(){
+        await actions.rFetchDatabases(this.props);
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/DatabaseListContainer.jsx]: again this strange `rFetch` pattern

File: frontend/src/metabase/reference/databases/FieldDetailContainer.jsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        table: PropTypes.object.isRequired,
+        field: PropTypes.object.isRequired,
+        isEditing: PropTypes.bool
+    };
+
+    async fetchContainerData(){
+        await actions.rFetchDatabaseMetadata(this.props, this.props.databaseId);
+    }
+
+    async componentWillMount() {
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/FieldDetailContainer.jsx]: this doesn't need to be async - applies to componentWillReceiveProps too

Comment (attekei) [inline frontend/src/metabase/reference/databases/FieldDetailContainer.jsx]: (React doesn't care at all about the return value of `componentWillMount`, `componentWillReceiveProps` etc)

File: frontend/src/metabase/reference/databases/FieldList.jsx [hunk 11]
```diff
@@ -39,17 +37,23 @@ import { getIconForField } from "metabase/lib/schema_metadata";
 import * as metadataActions from "metabase/redux/metadata";
 import * as actions from 'metabase/reference/reference';
 
+
+const emptyStateData = {
+            message: `Fields in this table will appear here as they're added`,
+            icon: "fields"
+        }
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/FieldList.jsx]: indentation pls – you could maybe run all data reference section files through `prettier` so that you don't have to care about formatting yourself (kyle is an expert on prettier stuff)

File: frontend/src/metabase/reference/databases/FieldListContainer.jsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        database: PropTypes.object.isRequired,
+        table: PropTypes.object.isRequired,
+        isEditing: PropTypes.bool
+    };
+
+    async fetchContainerData(){
+        await actions.rFetchDatabaseMetadata(this.props, this.props.databaseId);
+    }
+
+    async componentWillMount() {
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/FieldListContainer.jsx]: again unnecessary async

File: frontend/src/metabase/reference/databases/databases.integ.spec.js [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            expect(container.find(ReferenceHeader).length).toBe(1)
+            expect(container.find(DatabaseList).length).toBe(1)            
+            expect(container.find(AdminAwareEmptyState).length).toBe(0)
+            
+            expect(container.find(List).length).toBe(1)
+            expect(container.find(ListItem).length).toBe(1)
+        })
+        
+        // database detail
+        it("should see a the detail view for the sample database", async ()=>{
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/databases.integ.spec.js]: this should probably check that dom contains some expected information about the database?

File: frontend/src/metabase/reference/databases/databases.integ.spec.js [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+            expect(4).toBe(4);
+        })
+        // table detail
+
+       it("should see the Orders table", async  () => {
+            const store = await createTestStore()
+            store.pushPath("/reference/databases/1/tables/1");
+            mount(store.connectContainer(<TableDetailContainer />));
+            await store.waitForActions([FETCH_DATABASE_METADATA])
+            expect(true).toBe(true);
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/databases.integ.spec.js]: `expect(true).toBe(true);` is unnecessary – the test will pass if the code doesn't produce errors

File: frontend/src/metabase/reference/databases/databases.integ.spec.js [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+
+            expect(true).toBe(true);
+        })
+       it("should see the questions for the orders tables", async  () => {
+
+            const store = await createTestStore()
+            store.pushPath("/reference/databases/1/tables/1/questions");
+            mount(store.connectContainer(<TableQuestionsContainer />));
+            await store.waitForActions([FETCH_DATABASE_METADATA])
+            expect(true).toBe(true);
```
Comment (attekei) [inline frontend/src/metabase/reference/databases/databases.integ.spec.js]: `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);` `expect(true).toBe(true);`

File: frontend/src/metabase/reference/guide/guide.integ.spec.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        
+        it("Should show an empty guide for non-admin users", async () => {
+            const store = await createTestStore()    
+            store.pushPath("/reference/");
+            mount(store.connectContainer(<GettingStartedGuideContainer />));
+            await store.waitForActions([FETCH_DATABASE_METADATA, FETCH_SEGMENTS, FETCH_METRICS])
+        })
+        
+        it("Should show an empty guide with a creation CTA for admin users", async () => {
+            // TODO
```
Comment (attekei) [inline frontend/src/metabase/reference/guide/guide.integ.spec.js]: use `pending()` instead of letting the test stubs pass

File: frontend/src/metabase/reference/metrics/MetricListContainer.jsx [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        this.fetchContainerData()
+    }
+
+
+    async componentWillReceiveProps(newProps) {
+        if (this.props.location.pathname === newProps.location.pathname) {
+            return;
+        }
+
+        newProps.endEditing();
```
Comment (attekei) [inline frontend/src/metabase/reference/metrics/MetricListContainer.jsx]: theis action call cluster really, really needs to be in a single action as it is being repeated in many components

**Review by attekei — CHANGES_REQUESTED** [review]
Added some comments, should be relatively easy fixes! 

Lots and lots of repetition and boilerplate but of course it's still a major improvement to have each section in their component.

Found out that if you go to `/reference/segments/1` it throws the `Cannot read property 'table_id' of undefined` error. Interestingly for metrics everything works as intended.

File: frontend/src/metabase/reference/components/EditableReferenceHeader.jsx (hunk 5 above)
Comment (salsakran) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: The unwinding of combination forms + displays isn't part of this PR. 

I'll probably tackle it in a subsequent one once we've figured out what our form story is.

File: frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx (hunk 6 above)
Comment (salsakran) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: r is shorthand for "wrapped".

Some of the actions are "wrapped" (i.e. set loading + error statuses). I'm explicitly marking until this pattern has been resolved.

Imo, I think my plan for this refactor is:
1. Unwind the whole Reference* components (this PR)
2. Unwind the edit+display single component pattern
3. Unify the way fetches, etc are handled with the data model 

Separately, I'd like to reduce the number of things loaded, which I think can happen independently of 2 + 3.

Comment (camsaul) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: I second @attekei's comment. It's not obvious that `r` stands for `wrapped`. 

I think any time you have to explain what the meaning of the hungarian notation prefixes you're putting in front of your variables you should stop and consider whether introducing them is a good idea. A more verbose name like `wrappedFetchDatabaseMetadata` at least isn't going to have a bunch of potential contributors and future new hires scratching their head or putting `r` in from of random variables because they don't understand the pattern

File: frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx (hunk 7 above)
Comment (camsaul) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: can we do a [Fluent API](https://en.wikipedia.org/wiki/Fluent_interface)?

Comment (salsakran) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: that's a bit weird wrt how actions are normally bound and dispatched.

File: frontend/src/metabase/reference/components/EditableReferenceHeader.jsx (hunk 5 above)
Comment (camsaul) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: FWIW nested ternaries are great if you indent them properly. Then they become sort of like a `case` statement in `Clojure`. For example here's a bad FizzBuzz that at least has a nice ternary statement

```c
#include <stdio.h>

int main() {
  for (int i = 0; i < 100; i++) {
     printf((!(i % 15) ? "%d FizzBuzz\n" :
             !(i % 3)  ? "%d Fizz\n" :
             !(i % 5)  ? "%d Buzz\n" :
                         "%d\n"),
            i);
  }
  return 0;
}
```

it's preferable to a bunch of `if`s and `else`s I think

Comment (camsaul) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: Also FWIW try doing the same thing as my example with `if`s and `else`s. You can't in a language in the C-syntax family. The ternary operator is more flexible! 💯

File: frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx (hunk 6 above)
Comment (salsakran) [inline frontend/src/metabase/reference/databases/DatabaseDetailContainer.jsx]: points taken.

File: frontend/src/metabase/reference/components/EditableReferenceHeader.jsx (hunk 5 above)
Comment (salsakran) [inline frontend/src/metabase/reference/components/EditableReferenceHeader.jsx]: The main issue imo is not a ternary operator but the fact that something that imo should be two different components (an edit state and the normal read-only state) are conflated. Rather than two very simple components, each detail and list component has  ternary operators that effectively emit one of two components depending on the edit state.

**Comment by salsakran** [comment]
@attekei the foreign key, and edit state not carrying over old values are not regressions.
I fixed the direct link to a field.

File: frontend/src/metabase/reference/reference.js [hunk 18]
```diff
@@ -31,42 +49,499 @@ export const fetchGuide = createThunkAction(FETCH_GUIDE, (reload = false) => {
...
+                    [props.fetchSegmentFields(sID),
+                     props.fetchSegmentTable(sID)]
+                )}
+        )(segmentID)
+}
+
+// This is called when a component gets a new set of props.
+// I *think* this is un-necessary in all cases as we're using multiple 
+// components where the old code re-used the same component
+export const clearState = props => {
```
Comment (tlrobinson) [inline frontend/src/metabase/reference/reference.js]: It's slightly confusing that this function isn't actually an action creator but it's in the "duck" file along with other actions, and being called like `actions.clearState(newProps)`. It also feels a little weird to have any mention of "props" in Redux code.

We could make it an actual thunk action creator that dispatches each of the other actions. Or move it (and `resetForm` and any other similar fns) to another file. Or just leave it as is and document this minor weirdness.

**Review by attekei — APPROVED** [review]

**Comment by attekei** [comment]
Looks pretty good to me now! I agree with @tlrobinson:s comment about  `clearState` and `resetForm` don't follow normal redux patterns but that can be addressed during the next data reference refactoring iteration.

I was also going to say that would be good to run also e2e tests with `[ci e2e]` before merging the branch – let's see if they pass on master ;)
